### PR TITLE
PSA: Shared code between module-specific integration work

### PR DESCRIPTION
**Pull request by hanno-arm**
__Summary:__ This PR adds the header file `mbedtls/psa_util.h` which contains static utility functions `mbedtls_psa_xxx()` used in the integration of PSA Crypto into Mbed TLS. These functions are internal only and may change at any time.

This is likely to expand over time and so far contains only the functions introduced by @mpg and @AndrzejKurek in PRs #2162 and #2159.

@AndrzejKurek @mpg Please let me know what else you need and I'll add it here.

__Internal Reference:__ IOTSSL-2621.

### CODE REVIEW

File: include/mbedtls/psa_util.h [hunk 1]
```diff
@@ -0,0 +1,162 @@
+/**
+ * \file psa_compat.h
+ *
+ * \brief Utility functions for the use of the PSA Crypto library.
+ *
+ * \warning This function is not part of the public API and may
```
Comment (mpg) [inline include/mbedtls/psa_util.h]: Minor: "this file" or "these functions".

File: include/mbedtls/psa_util.h [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+#include "psa/crypto.h"
+
+#include "ecp.h"
+#include "md.h"
+
+static psa_status_t mbedtls_psa_get_free_key_slot( psa_key_slot_t *key )
+{
+    for( psa_key_slot_t slot = 1; slot <= 32; slot++ )
+    {
+        if( psa_get_key_information( slot, NULL, NULL )  == PSA_ERROR_EMPTY_SLOT )
```
Comment (mpg) [inline include/mbedtls/psa_util.h]: Suggestion for improvement: if `psa_get_key_information()` returns `PSA_ERROR_BAD_STATE`, immediately return that to the caller, instead of returning `PSA_ERROR_INSUFFICIENT_MEMORY` later. This would make it slightly easier to debug stupid mistakes such as forgetting to call `psa_crypto_init()`.

**Comment by mpg** [comment]
@hanno-arm @AndrzejKurek I think this PR might actually be a good place to add a call to `psa_crypto_init()` to the `main()` function in `main.function` (for the test suites) and to the example programs. Otherwise it's likely we'll end up adding in independently in at least `main.function` and the SSL test programs, resulting in merge conflict. Wdyt?

(This would rename the PR from "add header" to "add common things" or something like that.)

**Comment by hanno-arm** [comment]
@mpg Yes indeed, just did that a few minutes ago, will push the commit soon.

File: include/mbedtls/psa_util.h (hunk 1 above)
Comment (hanno-arm) [reply include/mbedtls/psa_util.h]: Fixing.

**Comment by mpg** [comment]
@hanno-arm Damn, you code faster than I can comment :)

File: include/mbedtls/psa_util.h [hunk 3]
```diff
@@ -0,0 +1,162 @@
+/**
+ * \file psa_compat.h
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/psa_util.h]: Fixed in a8858f0.

**Comment by hanno-arm** [comment]
@mpg @AndrzejKurek I added an error conversion function from #2162.

**Comment by hanno-arm** [comment]
@mpg @AndrzejKurek Sorry, there's a stupid typo which even prevents compilation. I'll fix immediately.

File: include/mbedtls/psa_util.h (hunk 1 above)
Comment (AndrzejKurek) [reply include/mbedtls/psa_util.h]: Up.

File: include/mbedtls/psa_util.h [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+}
+
+/* Translations for symmetric crypto. */
+
+static inline psa_key_type_t mbedtls_psa_translate_cipher_type(
+    mbedtls_cipher_type_t cipher )
+{
+    switch( cipher )
+    {
+        case MBEDTLS_CIPHER_AES_128_CCM:
```
Comment (AndrzejKurek) [inline include/mbedtls/psa_util.h]: Shouldn't we check for MBEDTLS_AES_C?

File: include/mbedtls/psa_util.h [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+{
+    switch( status )
+    {
+        case PSA_SUCCESS:
+            return( 0 );
+        case PSA_ERROR_NOT_SUPPORTED:
+            return( MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE );
+        case PSA_ERROR_INSUFFICIENT_MEMORY:
+            return( MBEDTLS_ERR_PK_ALLOC_FAILED );
+        case PSA_ERROR_COMMUNICATION_FAILURE:
```
Comment (AndrzejKurek) [inline include/mbedtls/psa_util.h]: I would add a comment about the fallthrough here.

File: tests/suites/main_test.function [hunk 6]
```diff
@@ -221,8 +224,22 @@ int main( int argc, const char *argv[] )
                          ret );
         return( -1 );
     }
+
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    {
```
Comment (AndrzejKurek) [inline tests/suites/main_test.function]: Why did you add the outer layer of curly braces?

Comment (hanno-arm) [reply tests/suites/main_test.function]: We must declare local variables at the beginning of their scope, otherwise the compilers complain.

Comment (mpg) [reply tests/suites/main_test.function]: Hopefully we can soon relax that restriction when we move to using more of C99, but currently we do indeed compile with `-Wdeclaration-after-statement` until (a larger subset of) C99 is officially declared acceptable for us to use.

Comment (AndrzejKurek) [reply tests/suites/main_test.function]: Sorry, I asked because I was curious about why didn't you opt for having the declarations at the beginning of the function, therefore - omitting these braces. But I guess that's just a matter of style.

**Comment by hanno-arm** [comment]
@mpg @AndrzejKurek I have created a branch `feature-psa` which is the merge of PSA's `feature-psa` with current `development`, and rebased @mpg's `psa-port-evaluation` and the commits from this PR on top of it.

File: include/mbedtls/psa_util.h (hunk 4 above)
Comment (hanno-arm) [reply include/mbedtls/psa_util.h]: I don't think so, because when PSA Crypto and Mbed TLS are separated, we shouldn't consider Mbed TLS' configuration options to decide whether to use a particular PSA identifier. Currently, we could do that, but I prefer to leave things as they are and to get the failure when attempting to setup the PSA context.

File: include/mbedtls/psa_util.h (hunk 5 above)
Comment (hanno-arm) [reply include/mbedtls/psa_util.h]: I pushed 359f62 to improve the documentation here.

**Comment by hanno-arm** [comment]
@AndrzejKurek Thank you for your comments, I believe I have addressed them. Could you please re-review?

**Comment by hanno-arm** [comment]
@mpg @AndrzejKurek @sbutcher-arm As discussed, the utility branch `psa-integration-utilities` is temporarily becoming the base for the module-specific PSA integration PRs, replacing `psa-port-evaluation`. For that, I had to create it as a branch in this repository, so the source `hanno-arm/psa-integration-utilities` of this PR is out of date. Closing this PR in favour of a new one which uses the `mbedtls/psa-integration-utilities` as the source.
